README.markdown
[![Build Status](https://travis-ci.org/olivierHa/puppet-influxdb.svg)](https://travis-ci.org/olivierHa/puppet-influxdb)

####Table of Contents

1. [Overview](#overview)
2. [Module Description - What the module does and why it is useful](#module-description)
3. [Setup - The basics of getting started with influxdb](#setup)
    * [What influxdb affects](#what-influxdb-affects)
    * [Setup requirements](#setup-requirements)
    * [Beginning with influxdb](#beginning-with-influxdb)
4. [Usage - Configuration options and additional functionality](#usage)
5. [Reference - An under-the-hood peek at what the module is doing and how](#reference)
5. [Limitations - OS compatibility, etc.](#limitations)
6. [Development - Guide for contributing to the module](#development)

##Overview

The influxdb module lets you use Puppet to install, configure, and manage Influxdb version 0.10.x and 0.9.x
The current InfluxDB version is 0.10

##Module Description

InfluxDB is a Open source distributed time series, events, and metrics database. 
This module handles the installation  and every configuration parameters.

Support for Puppet 3.x and Puppet 4.x

##Setup

###Beginning with influxdb

To install InfluxDB with the default parameters

~~~puppet
  class { 'influxdb': }
~~~

###What influxdb affects

* influxdb package.
* influxdb configuration file.
* influxdb service.

##Usage

To enable Graphite plugin : 

~~~puppet
  class { 'influxdb': 
    graphite => {
     main => {
        enabled => true,
     }
    }
  }
~~~

With InfluxDB, you can configure several graphite backends.

####A more complex example

~~~puppet
  class { 'influxdb': 
    graphite => {
     main => {
        enabled => true,
        database      => 'mygraphitedb',
        batch_size    => 500,
        tags          => ['region=us-east', 'zone=1c'],
        templates     => [
           "servers.* .host.measurement*",
           "stats.* .host.measurement* region=us-west,agent=sensu",
        ]
     },
     udp => {
        enabled => true,
        bind-address => ':20003',
        protocol => 'udp',
     }
    }
  }
~~~

To enable Collectd plugin:

~~~puppet
  class { 'influxdb': 
    collectd => {  
      enabled => true,
    }
  }
~~~

####Hiera example

```yaml
influxdb::package_ensure: 'latest'
influxdb::graphite:
 main:
   enabled: true
   batch_size: 3
influxdb::data:
  dir: '/mnt/influxdb/data'
```

##Reference

###Classes

####Public Classes

* [`influxdb`](#class-influxdb): Guides the basic setup of InfluxDB.

####Private Classes

* `influxdb::install`: Installs InfluxDB packages.
* `influxdb::config`: Configures InfluxDB.
* `influxdb::params`: Manages InfluxDB parameters.
* `influxdb::service`: Manages the InfluxDB daemon.

###Parameters

As InfluxDB is moving quickly, this module has a "dynamic" structure.
Each section of the Influxdb configuration file is mapped to a hash.
So you can easily add/remove new parameters as the influxdb software evolves, without change this module.

####`Section Hashes`

Each config file section accepts a hash of values. Defaults are set in params.pp

Here are the default values for the meta section :

~~~puppet

  $meta = {
    dir => '/var/lib/influxdb/meta',
    hostname => 'localhost',
    bind-address => ':8088',
    retention-autocreate => true,
    election-timeout => '1s',
    heartbeat-timeout => '1s',
    leader-lease-timeout => '500ms',
    commit-timeout => '50ms',
    cluster-tracing => false,
    raft-promotion-enabled => true,
    logging-enabled => true,
  }
~~~

####`package_ensure`
Choose whether to install or uninstall, or force to the latest version. Default: present

####`manage_repo`
Choose whether to configure influxdb repositories. Default: true

####`config_file`
Full path to config file (e.g. /etc/influxdb/influxdb.conf). Defaults to "/etc/influxdb/influxdb.conf"

####`influxdb_user`
Ownership of influxdb directories. (Don't set user that will run influxdb ... yet) Default: 'influxdb'

####`influxdb_group`
Group of influxdb directories. Default: 'influxdb'

####`conf_template`
Specify template to use for influxdb conf file. Defaults to "influxdb/influxdb.conf.erb"

###Defined Types

#### influxdb_database

~~~
influxdb_database { 'graphite':
  ensure  => 'present',
}
~~~

#### influxdb_retention_policy

`influxdb_retention_policy` creates retention policy on databases. 
To use it you must create the title of the resource as shown below,
following the pattern of `policy_name@database`:

~~~
influxdb_retention_policy { '1w@graphite':
  ensure      => present,
  database    => 'graphite',
  duration    => '1w',
  is_default  => true,
  replication => '1',
}
~~~

##Limitations

This module has been tested on:

 - Debian 7 Wheezy
 - Debian 8 Jessie

It **should** also work on Debian-based OS (Ubuntu) and RedHat osfamily

This module is fully compatible with Puppet 3.x and Puppet 4.x

Providers work without authentication/authorization (work in progress)

## Contributing

Please report bugs and feature request using [GitHub issue
tracker](https://github.com/olivierHa/puppet-influxdb/issues)

##Development

 - rspec and beaker tests
 - authentication/authorization for provider


CHANGELOG
2016-02-05 Release 0.5.0
- Update to Influxdb 0.10
- Refactoring of the module. Now every section is hash that can be merged with a custom hash. 
- Support of multiple Graphite sections

2015-08-27 Release 0.4.2
- Influxdb 0.9.3

2015-08-14 Release 0.4.1
- Better handling of duration format for Retention Policy type/provider

2015-08-12 Release 0.4.0
- Provider and Type for Database and Retention Policy (Beta)
- Fix typo in configuration template about graphite database
- Update default graphite batch size

2015-08-01 Release 0.3.1
- Fix bug with config_file on Redhat osfamily
- Add some basic unit tests

2015-07-29 Release 0.3.0
- Handle new parameters : admin, http, opentsdb, Shard precreation
- More documentation

2015-07-28 Release 0.2.0
- Handle new parameters : collectd, monitoring, Hinted-handoff and Continuous_queries
- Handle Clustering

2015-07-27 Release 0.1.1
- Better metadata.json for puppetforge

2015-07-27 Release 0.1.0
- Initial Release (Basic installation and graphite support)


spec/classes/example_spec.rb
require 'spec_helper'

describe 'influxdb' do
  context 'supported operating systems' do
    ['Debian', 'RedHat'].each do |osfamily|
      describe "influxdb class without any parameters on #{osfamily}" do
        let(:params) {{ }}
        let(:facts) {{
          :osfamily => osfamily,
        }}

        it { should compile.with_all_deps }

        it { should contain_class('influxdb::params') }
        it { should contain_class('influxdb::install').that_comes_before('influxdb::config') }
        it { should contain_class('influxdb::config') }
        it { should contain_class('influxdb::service').that_subscribes_to('influxdb::config') }

        it { should contain_service('influxdb') }
        it { should contain_package('influxdb').with_ensure('present') }
      end
    end
  end

  context 'unsupported operating system' do
    describe 'influxdb class without any parameters on Solaris/Nexenta' do
      let(:facts) {{
        :osfamily        => 'Solaris',
        :operatingsystem => 'Nexenta',
      }}

      it { expect { should contain_package('influxdb') }.to raise_error(Puppet::Error, /Nexenta not supported/) }
    end
  end
end


spec/classes/install_spec.rb
require 'spec_helper'

describe 'influxdb', :type => :class do
  context "defaut install parameters" do
    let :facts do
      {
        :osfamily => 'Debian',
        :lsbdistid => 'Debian',
        :operatingsystem => 'Debian',
        :operatingsystemrelease => '7.8',
        :lsbdistcodename => 'wheezy',
      }
    end
   it do
     should contain_package('influxdb').with_ensure('present')
   end
  end
end


spec/classes/service_spec.rb
require 'spec_helper'

describe 'influxdb::service', :type => :class do
  let :pre_condition do
    'include influxdb::params'
  end
  context "on a Debian OS" do
    let :facts do
      {
        :osfamily               => 'Debian',
        :operatingsystemrelease => '7',
        :concat_basedir         => '/dne',
        :lsbdistid              => 'Debian',
        :lsbdistcodename        => 'wheezy',
        :operatingsystem        => 'Debian',
        :id                     => 'root',
        :kernel                 => 'Linux',
        :path                   => '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
        :is_pe                  => false,
      }
    end

    it { is_expected.to contain_service("influxdb").with(
      'name'      => 'influxdb',
      'ensure'    => 'running',
      'enable'    => 'true'
      )
    }

    context "with $service_name => 'foo'" do
      let (:params) {{ :service_name => 'foo' }}
      it { is_expected.to contain_service("influxdb").with(
        'name'      => 'foo'
        )
      }
    end

    context "with $service_enable => true" do
      let (:params) {{ :service_enable => true }}
      it { is_expected.to contain_service("influxdb").with(
        'name'      => 'influxdb',
        'ensure'    => 'running',
        'enable'    => 'true'
        )
      }
    end

    context "with $service_enable => false" do
      let (:params) {{ :service_enable => false }}
      it { is_expected.to contain_service("influxdb").with(
        'name'      => 'influxdb',
        'ensure'    => 'running',
        'enable'    => 'false'
        )
      }
    end

    context "$service_enable must be a bool" do
      let (:params) {{ :service_enable => 'not-a-boolean' }}

      it 'should fail' do
        expect { catalogue }.to raise_error(Puppet::Error, /is not a boolean/)
      end
    end

    context "with $service_ensure => 'running'" do
      let (:params) {{ :service_ensure => 'running', }}
      it { is_expected.to contain_service("influxdb").with(
        'ensure'    => 'running',
        'enable'    => 'true'
        )
      }
    end

    context "with $service_ensure => 'stopped'" do
      let (:params) {{ :service_ensure => 'stopped', }}
      it { is_expected.to contain_service("influxdb").with(
        'ensure'    => 'stopped',
        'enable'    => 'true'
        )
      }
    end

    context "with $service_ensure => 'UNDEF'" do
      let (:params) {{ :service_ensure => 'UNDEF' }}
      it { is_expected.to contain_service("influxdb").without_ensure }
    end

  end

  context "on a RedHat 6 OS, do not manage service" do
    let :facts do
      {
        :osfamily               => 'RedHat',
        :operatingsystemrelease => '6',
        :concat_basedir         => '/dne',
        :operatingsystem        => 'RedHat',
        :lsbdistid              => 'RedHat',
        :id                     => 'root',
        :kernel                 => 'Linux',
        :path                   => '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
        :is_pe                  => false,
      }
    end
    let(:params) do
      {
        'service_ensure' => 'running',
        'service_name'   => 'influxdb',
        'service_manage' => false
      }
    end
    it 'should not manage the influxdb service' do
      expect(subject).not_to contain_service('influxdb')
    end
  end

end


lib/puppet/provider/influxdb_retention_policy/ruby.rb
Puppet::Type.type(:influxdb_retention_policy).provide(:ruby) do
  desc 'Manages InfluxDB retention policy.'

commands :influx_cli => '/usr/bin/influx'

  mk_resource_methods

  def initialize(value={})
    super(value)
    @property_flush = {}
  end

  def self.instances
    instances = []
    begin
      dbs = influx_cli(['-execute', 'show databases'])
    rescue Puppet::ExecutionFailure => e
      Puppet.debug("#show databases had an error -> #{e.inspect}")
      return nil
    end
    databases = dbs.split("\n")[3..-1]

    databases.collect do |db|
      begin 
        output = influx_cli(['-execute', "show retention policies on #{db}"].compact)
      rescue Puppet::ExecutionFailure => e
        Puppet.debug("#show retention policies had an error -> #{e.inspect}")
        return nil
      end
      policies = output.split("\n")[1..-1]
      policies.each do |policy|
        ret_name , duration , replica, is_default = policy.split("\t").reject { |e| e.to_s.empty? }
        dur_s = self.duration_to_s(duration)
        instances << new(
           :name        => "#{ret_name}@#{db}",
           :ensure      => :present,
           :database    => db,
           :replication => replica,
           :duration    => dur_s,
           :is_default  => is_default
          )
      end
    end

  return instances
  end

  def self.prefetch(resources)
    instances.each do |prov|
      if resource = resources[prov.name]
        resource.provider = prov
      end
    end
  end

  def flush
    if @property_flush[:ensure] == :absent
        influx_cli(['-execute', "drop RETENTION POLICY #{resource[:name]} ON #{resource[:database]}"].compact)
        return
    end
    short_ret_name = "#{resource[:name]}".split('@').first
    if @property_flush[:ensure] == :present
      if resource[:is_default]
        influx_cli(['-execute', "create RETENTION POLICY \"#{short_ret_name}\" ON #{resource[:database]} DURATION #{resource[:duration]} REPLICATION #{resource[:replication]} DEFAULT"].compact)
      else
        influx_cli(['-execute', "create RETENTION POLICY \"#{short_ret_name}\" ON #{resource[:database]} DURATION #{resource[:duration]} REPLICATION #{resource[:replication]}"].compact)
      end
      return
    end
    cmd_arguments = "ALTER RETENTION POLICY \"#{short_ret_name}\" ON #{resource[:database]} "
    if @property_flush and ! @property_flush.empty?
      cmd_arguments << ' DURATION ' << @property_flush[:duration] if @property_flush[:duration]
      cmd_arguments << ' REPLICATION ' << @property_flush[:replication] if @property_flush[:replication]
      cmd_arguments << ' DEFAULT' if @property_flush[:is_default]
      influx_cli(['-execute', "#{cmd_arguments}"].compact)
    end
  end

  def exists?
    @property_hash[:ensure] == :present || false
  end

  def create
    @property_flush[:ensure] = :present
  end

  def destroy
    @property_flush[:ensure] = :absent
  end

  def self.duration_to_s(dur)
    re = /^(\d+w)?(\d+d)?(\d+h)?(\d+m)?(\d+s)?$/
    match = dur.match re
    if match
      val = 0
      val += match[1].chop.to_i * 604800 if match[1]
      val += match[2].chop.to_i * 86400 if match[2]
      val += match[3].chop.to_i * 3600 if match[3]
      val += match[4].chop.to_i * 60 if match[4]
      val += match[5].chop.to_i if match[5]
      return "#{val}s"
    end
    return nil
  end

  def duration=(duration)
    d1 = self.class.duration_to_s(duration)
    d2 = self.class.duration_to_s(@property_hash[:duration])
    if !d1.nil? and !d2.nil? and d1.to_i != d2.to_i
      @property_flush[:duration] = duration
    end
  end

  def replication=(replication)
    @property_flush[:replication] = replication
  end

  def is_default=(is_default)
    if is_default.to_s == 'true'
      @property_flush[:is_default] = is_default
    end
  end

end


lib/puppet/type/influxdb_retention_policy.rb
Puppet::Type.newtype(:influxdb_retention_policy) do
  @doc = 'Manage Influxdb retention policies.'

  ensurable

  autorequire(:class) { 'influxdb::install' }

  newparam(:name, :namevar => true) do
    desc 'The name of the retention to manage.'
  end

  def munge_boolean(value)
    case value
    when true, "true", :true
      :true
    when false, "false", :false
      :false
    else
      fail("munge_boolean only takes booleans")
    end
  end

  def duration_to_s(dur)
    re = /^(\d+w)?(\d+d)?(\d+h)?(\d+m)?(\d+s)?$/
    match = dur.match re
    if match
      val = 0
      val += match[1].chop.to_i * 604800 if match[1]
      val += match[2].chop.to_i * 86400 if match[2]
      val += match[3].chop.to_i * 3600 if match[3]
      val += match[4].chop.to_i * 60 if match[4]
      val += match[5].chop.to_i if match[5]
      return "#{val}s"
    end
    return nil
  end

  newproperty(:database,) do
    desc "Database retention"
    newvalues(/^\S+$/)
  end

  newproperty(:duration,) do
    desc "Duration of retention, format is <INT>w<INT>d<INT>h<INT>s"
    newvalues(/^(\d+w)?(\d+d)?(\d+h)?(\d+m)?(\d+s)?$/)

    munge do |value|
      @resource.duration_to_s(value)
    end

  end
 
  newproperty(:replication,) do
    desc "Number of replication"
    newvalues(/^\d+$/)
  end

  newproperty(:is_default, :boolean => true) do
    desc "Is it the default retention policy for this database"
    newvalue(:true)
    newvalue(:false)
    munge do |value|
      @resource.munge_boolean(value)
    end
  end

end
